capsules/src/process_console.rs

```diff
@@ -945,6 +968,12 @@ impl<'a, A: Alarm<'a>, C: ProcessManagementCapability> uart::ReceiveClient
                 0 => debug!("ProcessConsole had read of 0 bytes"),
                 1 => {
                     self.command_buffer.map(|command| {
+                        if command[0] == 0 {
+                            //None is seen as a "user is typing" state
+                            //from the perspective of the command browser
+                            //and thus it returns to the default state
+                            self.command_history_index.insert(None);
```

Comment:
The way it is implemented right now, if the user has written any number of characters such as "pr" and accidentally presses an upper arrow the text the text will be overwritten with the previous command.

Two possible other desirable behaviours would be:
- pressing up arrow will do nothing
- the command history will display commands that start with the same string

Reply:
The "do nothing" idea is most at hand in terms of time. However, a radix trie or string BST could also resolve the second idea. Your suggestion ?

Reply:
> The "do nothing" idea is most at hand in terms of time. However, a radix trie or string BST could also resolve the second idea. Your suggestion ?

Considering that you have a list of max 100 commands, anything else than a list is not useful due to code size.

Reply:
Another suggestion is to memorize as the *last command* what the user just wrote, and if he presses down it will put the user's input back.

Reply:
I will pick this up with your idea after @Matrix22 is done restructuring the command history.

Reply:
@valexandru  It is done. I sincerely hope great joy finds you upon hearing this news